source/extensions/filters/http/language/language_filter.cc

```diff
@@ -0,0 +1,133 @@
+#include "source/extensions/filters/http/language/language_filter.h"
+
+#include <string>
+
+#include "envoy/common/exception.h"
+#include "envoy/extensions/filters/http/language/v3/language.pb.h"
+
+#include "source/common/common/empty_string.h"
+#include "source/common/common/logger.h"
+#include "source/common/common/utility.h"
+#include "source/common/http/utility.h"
+
+#include "unicode/locid.h"
+#include "unicode/utypes.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Language {
+
+LanguageFilterConfigImpl::LanguageFilterConfigImpl(
+    const envoy::extensions::filters::http::language::v3::Language& proto_config,
+    const std::string& stats_prefix, Stats::Scope& scope)
+    : stats_(LanguageFilter::generateStats(stats_prefix, scope)),
+      default_language_(proto_config.default_language().data()) {
+  for (const auto& supported_language : proto_config.supported_languages()) {
+    supported_languages_.push_back(icu::Locale(supported_language.data()));
+  }
+}
+
+LanguageStats& LanguageFilterConfigImpl::stats() { return stats_; }
+
+const icu::Locale& LanguageFilterConfigImpl::defaultLanguage() const { return default_language_; }
+
+const std::vector<icu::Locale>& LanguageFilterConfigImpl::supportedLanguages() const {
+  return supported_languages_;
+}
+
+LanguageFilter::LanguageFilter(const LanguageFilterConfigSharedPtr config)
+    : config_(std::move(config)) {
+  UErrorCode errorCode = U_ZERO_ERROR;
+
+  locale_matcher_ = std::make_shared<icu::LocaleMatcher>(
+      icu::LocaleMatcher::Builder()
+          .setSupportedLocales(config_->supportedLanguages().begin(),
+                               config_->supportedLanguages().end())
+          .setDefaultLocale(&config_->defaultLanguage())
+          .build(errorCode));
+
+  if (U_FAILURE(errorCode)) {
+    ENVOY_LOG(error, "ICU error code icu::LocaleMatcher::Builder: {}", errorCode);
+
+    throw EnvoyException("Failed to initialize icu::LocaleMatcher.");
```

Comment:
You can not throw from the filter constructor as we avoid using exceptions on the dataplane threads. You will need to create this matcher eiother as part of the factory or configuration, so that an error and exception can be handled during config processing.

Reply:
I did not know about that. Moved the code to the factory implementation.